### PR TITLE
Porting changes from OldDDerivations, attempt 2.

### PR DESCRIPTION
**Pull request by dukc**
~~Note: Draft since some C++ test is still failing at late phase (testing 2.102, NixOS with the current stable channel)~~

This supersedes #2

I have pulled the changes done to this repo since the original PR and reimplemented my changes on top of the latest main version.

@PetarKirov please do something this time once I get this out of draft. I do understand you want to implement the automatic hash fetcher script that will make the per-version nix files needless. I know your intention was to rebase my original MR yourself, which would solve the problem if waiting for it wasn't blocking me from doing anything else in the meantime.

However, compiler versions and NixPkgs changes keep coming, and I don't want to be blocked for months before dealing with them. In the present, I'd have to push updates to OldDDerivations with knowledge that I'll have to do them again when it comes the time to port them here. This is just unworkable - I should have to do my job only once. I think there are three possibilities:

 - Give me merge rights to this repo.
 - Merge (or request changes) to my PRs within a week or so.
 - Or, on the occasion you can't do that, don't make too much other changes before dealing with the PR so I don't have to redo it (or further PRs built on top of it).

### CODE REVIEW

**Comment by dukc** [comment]
Ping @PetarKirov this now works, ready for review.

**Comment by PetarKirov** [comment]
Hi @dukc, thanks for this PR (and sorry for causing you the trouble of having to redo #2). I'll try to finish reviewing it today.

Btw, thanks for the ping, I just saw it by chance. In the future, can you try o reach out to me on either Slack or Discord, so you don't end up waiting? In general, I am very rarely able to keep up with GH notifications.

**Comment by PetarKirov** [comment]
In the mean time, could you review #3 (commit by commit, as looking at the total diff would be quite difficult to follow). I think many of the changes there should be fine in isolation.

File: pkgs/dmd/generic.nix
Comment (PetarKirov) [inline pkgs/dmd/generic.nix:91]: Good catch, I wasn't aware of that 👍

File: pkgs/dmd/generic.nix
Comment (PetarKirov) [inline pkgs/dmd/generic.nix:207]: Thanks for investigating this test failure as well 👍

File: pkgs/dmd/binary.nix
Comment (PetarKirov) [inline pkgs/dmd/binary.nix:42]: I would prefer if we use this commit:
https://github.com/PetarKirov/dlang-nix/commit/0736584ceab34e1c3876a6e9b69e30fe750dc121 where I tried to provide more context about why this change is necessary.

File: pkgs/dmd/generic.nix
Comment (PetarKirov) [inline pkgs/dmd/generic.nix:81]: Nice 👍

File: pkgs/dmd/generic.nix
Comment (PetarKirov) [inline pkgs/dmd/generic.nix:130]: Nice 👍

File: pkgs/dmd/default.nix
Comment (PetarKirov) [inline pkgs/dmd/default.nix:1]: Nice 👍

**Review by PetarKirov — APPROVED** [review]
LGTM. There 3 things I'd like to do before merging:

* cherry-pick my commit for libgcc from #3 
* set up CI pipeline - again from #3, but I would need to update the actions to the most recent version
* reword your commits to follow the [conventional commits style](https://www.conventionalcommits.org/en/v1.0.0/#summary), as in the future I'd like to set up something like [semantic release](https://semantic-release.gitbook.io/semantic-release/)

**Comment by dukc** [comment]
> LGTM. There 3 things I'd like to do before merging: [snip]

I reworded the commits and changed the Libgcc dependency.

But I don't wish to include (or wait for) a CI in this PR. I very much do want a CI, but it's the sort of thing that should be done independently in a separate PR. Your existing CI work should need no adaptations for rebasing on this one, because the command for building default.nix is not changing.

**Comment by dukc** [comment]
In the future, can you try o reach out to me on either Slack or Discord, so you don't end up waiting?

Oh I'll try to keep in mind. Maybe I'll still ping too but when I really want attention then Slack it is.

**Comment by dukc** [comment]
> In the mean time, could you review https://github.com/PetarKirov/dlang-nix/pull/3 (commit by commit, as looking at the total diff would be quite difficult to follow). I think many of the changes there should be fine in isolation.

I will, at some point.

**Comment by PetarKirov** [comment]
I just came back from travel abroad and found a few spare minutes to fix the issues (not using `${dmdPrefix}` everywhere and attempting to delete `runnable_cxx/cppa.d`).

**Review by PetarKirov — DISMISSED** [review]
With the rebase done and the post-rebase issues fixed, everything looks good to me. Thanks a lot for contributing these changes, much appreciated!

**Review by PetarKirov — APPROVED** [review]

**Comment by PetarKirov** [comment]
@dukc I had to downgrade the default DMD version back to 2.100.2, as 2.102.2 didn't pass the macOS CI.

I have logged #8 to track this issue.

**Pull request merged**
